crates/shared/src/price_estimation/trade_finder.rs

```diff
@@ -86,3 +217,161 @@ impl From<TradeError> for PriceEstimationError {
         }
     }
 }
+
+/// Output of `Trader::settle` smart contract call.
+#[derive(Debug, Eq, PartialEq)]
+struct SettleOutput {
+    gas_used: U256,
+    trader_balances: Vec<I256>,
+    settlement_balances: Vec<I256>,
+}
+
+impl SettleOutput {
+    fn decode(output: &[u8]) -> Result<Self> {
+        let function = Trader::raw_contract().abi.function("settle").unwrap();
+        let tokens = function.decode_output(output)?;
+        let (gas_used, trader_balances, settlement_balances) =
+            Tokenize::from_token(Token::Tuple(tokens))?;
+        Ok(Self {
+            gas_used,
+            trader_balances,
+            settlement_balances,
+        })
+    }
+
+    fn trader_amounts(&self) -> Option<(U256, U256)> {
+        Some((
+            self.trader_balances.get(0)?.wrapping_abs().into_raw(),
+            self.trader_balances.last()?.into_raw(),
+        ))
+    }
```

Comment:
Should this fail if `self.trader_balances.len() == 1`?
Also to be more consistent you could use `self.trader_balances.first()` instead of `.get(0)`. Same comments apply to `Self::exexuted_amounts()` as well.

Reply:
I'm one of the [lucky 10000 today](https://xkcd.com/1053/)!

Reply:
So what about the case when `self.trader_balances.len() == 1`? I suppose the trade simulation contract technically supports that case (only 1 token and price in the inputs) but it seems to me like it shouldn't, no? It would mean you want to simulate a trade from token A to token B which seems nonsensical to me.